### PR TITLE
generateRandomString function

### PR DESCRIPTION
**Pull request by aaronli03**
## What
- added function generateRandomString
- added test for generateRandomString

## Why
- random string to be passed in as state in the Spotify auth flow
- test checks that output matches desired length and are unique

## Impact
- Will customer experience be significantly altered? No
- Does the code change cover more than one task? No

## Testing
- tested locally by running `npm run dev`

### CODE REVIEW

File: tests/api/auth.test.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { describe, it, expect } from 'vitest';
+import { generateRandomString } from '../../src/api/auth';
+
+describe('Auth', () => {
+	it('generateRandom String', () => {
+		const length = 16;
+		const randomString1 = generateRandomString(length);
+		const randomString2 = generateRandomString(length);
+		expect(randomString1).not.toBe(randomString2);
+		expect(randomString1.length).toBe(randomString2.length);
```
Comment (byjackli) [inline tests/api/auth.test.ts]: We know that randomString1 is the same length as randomString2, but do we know that they're the expected length of..? 16? 32? 64? 42069?

File: src/api/auth.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export function generateRandomString(length: number) {
+	let result = '';
+	const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
+	for (let i = 0; i < length; i++) {
+		result += possible.charAt(Math.floor(Math.random() * possible.length));
+	}
+	return result;
```
Comment (byjackli) [inline src/api/auth.ts]: Is this Spotify's own unique way of wanting developers to generate a random string?

If not maybe try exploring something in like Crypto?
https://developer.mozilla.org/en-US/docs/Web/API/Crypto

**Review by byjackli — CHANGES_REQUESTED** [review]

File: tests/api/auth.test.ts (hunk 1 above)
Comment (aaronli03) [reply tests/api/auth.test.ts]: that was a typo, will update to check that the resulting length matches desired length

File: src/api/auth.ts (hunk 2 above)
Comment (aaronli03) [reply src/api/auth.ts]: This method was taken from Spotify's own implementation example on their document:
https://github.com/spotify/web-api-examples/blob/master/authentication/authorization_code/app.js
I also noticed I had a small typo in the allowed characters which is used for the code verifier and not here.
Using crpyto will add a tiny bit of complexity but will achieve a cryptographically secure source of random values.

Comment (byjackli) [reply src/api/auth.ts]: CryptoUUID and substring?

File: src/api/auth.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import * as crypto from 'crypto';
+
+export function generateRandomString(length: number) {
+	const characterSet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
```
Comment (byjackli) [inline src/api/auth.ts:4]: Better to make this as an `export const CHARACTER_STE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789" as const`, then import it from your `auth.test.ts` file

**Review by byjackli — APPROVED** [review]

**Review by byjackli — APPROVED** [review]
Better to make this as an `export const CHARACTER_STE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789" as const`, so you can also import it in `auth.test.ts` file

**Pull request merged**